### ISSUE TITLE
mqtt_client_dowork crash due to holding reference to destroyed xio

### ISSUE THREAD

**Issue by ccli8** [Azure/azure-iot-sdk-c#1666]
**Development Machine, OS, Compiler (and Other Relevant Toolchain Info)**

Mbed OS: 6.3.0
Toolchain: Arm Compiler 6.13

Arm Mbed ported Azure example:
https://github.com/ARMmbed/mbed-os-example-for-azure

**SDK Version (Please Give Commit SHA if Manually Compiling)**

**azure-iot-sdk-c**
LTS_07_2020_Ref01

**azure-umqtt-c**
2020-07-19

**Protocol**

MQTT

**Describe the Bug**

`mqtt_client.c` can hold `xio` object, which has been destroyed in `iothubtransportmqtt.c`. Please follow the sequence to replay it:

1. On mqtt disconnect, `DisconnectFromClient` gets called and `xio` object is destroyed:

    https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L2407-L2412

    https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L1860-L1864

    https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L298-L302

1. `IoTHubTransport_MQTT_Common_DoWork` still rolls, and invokes `mqtt_client_dowork`:

    https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L3131-L3134
    https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L3228-L3229

1. In `mqtt_client_dowork`, program crash in `xio_dowork` because `mqtt_client->xioHandle` has been destroyed by above:

    https://github.com/Azure/azure-umqtt-c/blob/165f6f52f8b2aec54a28ea5ede41736589726cd0/src/mqtt_client.c#L1280-L1285

    https://github.com/Azure/azure-umqtt-c/blob/165f6f52f8b2aec54a28ea5ede41736589726cd0/src/mqtt_client.c#L1293-L1302

**Comment by danewalton** [comment]
Hi @ccli8 quick question just to understand the situation a little better.

Are you attempting to disconnect from the endpoint and then call `do_work` after that?

**Comment by ccli8** [comment]
It seems I meet `IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN` error first and then go into mqtt disconnect sequence:

https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L1901-L1907

https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L2407-L2412

**Comment by danewalton** [comment]
I see are you able to post full logs with 
```c
bool traceOn = true;
IoTHubDeviceClient_LL_SetOption(device_ll_handle, OPTION_LOG_TRACE, &traceOn);
```

**Comment by ccli8** [comment]
The example uses `IoTHubDeviceClient_CreateFromConnectionString`, so I change to `IoTHubDeviceClient_SetOption`:

https://github.com/ARMmbed/mbed-os-example-for-azure/blob/7b4116b0392e41b7ae0cdd7faaa7ee0d36e630a7/main.cpp#L76-L79

**Log with crash**:

```
PHY ID 1:0x243
PHY ID 2:0xc54
Info: Connection success, MAC: 96:38:0:53:80:5c
Info: Getting time from the NTP server
Info: Time: Fri Sep25 2:35:58 2020

Info: RTC reports Fri Sep25 2:35:58 2020

Info: Starting the Demo
Info: Initializing IoT Hub client
Info: Sending: "10 messages left to send, or until we receive a reply"
-> 2:36:0 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: nuvoton-test-001.azure-devices.net/my-symm-device-001/?api-version=2019-10-01&DeviceClientType=iothubclient%2f1.3.9%20(native%3b%20mbedOS5%3b%20undefined) | PWD: XXXX | CLEAN: 0
<- 2:36:1 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
Info: Connected to IoT Hub
-> 2:36:1 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: devices/my-symm-device-001/messages/devicebound/# | QOS: 1
-> 2:36:1 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 3 | PAYLOAD_LEN: 53
<- 2:36:1 SUBACK | PACKET_ID: 2 | RETURN_CODE: 1
Info: Sending: "9 messages left to send, or until we receive a reply"
-> 2:36:1 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 4 | PAYLOAD_LEN: 52
<- 2:36:1 PUBACK | PACKET_ID: 3
Info: Message sent successfully
<- 2:36:2 PUBACK | PACKET_ID: 4
Info: Message sent successfully
Info: Sending: "8 messages left to send, or until we receive a reply"
-> 2:36:2 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 5 | PAYLOAD_LEN: 52
<- 2:36:3 PUBACK | PACKET_ID: 5
Info: Message sent successfully
Info: Sending: "7 messages left to send, or until we receive a reply"
-> 2:36:3 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 6 | PAYLOAD_LEN: 52
<- 2:36:4 PUBACK | PACKET_ID: 6
Info: Message sent successfully
Info: Sending: "6 messages left to send, or until we receive a reply"
-> 2:36:4 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 7 | PAYLOAD_LEN: 52
<- 2:36:5 PUBACK | PACKET_ID: 7
Info: Message sent successfully
Info: Sending: "5 messages left to send, or until we receive a reply"
-> 2:36:5 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 8 | PAYLOAD_LEN: 52
<- 2:36:6 PUBACK | PACKET_ID: 8
Info: Message sent successfully
Info: Sending: "4 messages left to send, or until we receive a reply"
-> 2:36:6 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 9 | PAYLOAD_LEN: 52
<- 2:36:7 PUBACK | PACKET_ID: 9
Info: Message sent successfully
Info: Sending: "3 messages left to send, or until we receive a reply"
-> 2:36:7 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 10 | PAYLOAD_LEN: 52
<- 2:36:8 PUBACK | PACKET_ID: 10
Info: Message sent successfully
Info: Sending: "2 messages left to send, or until we receive a reply"
-> 2:36:8 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 11 | PAYLOAD_LEN: 52
<- 2:36:9 PUBACK | PACKET_ID: 11
Info: Message sent successfully
Info: Sending: "1 messages left to send, or until we receive a reply"
-> 2:36:9 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 12 | PAYLOAD_LEN: 52
<- 2:36:10 PUBACK | PACKET_ID: 12
Info: Message sent successfully
-> 2:36:10 DISCONNECT

++ MbedOS Fault Handler ++

FaultType: HardFault

Context:
R0: 0
R1: 1C6822D1
R2: FFFFFFFC
R3: E10
R4: 2000BB10
R5: E10
R6: 0
R7: 40B0BF00
R8: 2000B830
R9: 1
R10: E0D0
R11: 2000B958
R12: 0
SP   : 20015788
LR   : 3C76F
PC   : 1C6822D0
xPSR : 1000000
PSP  : 20015720
MSP  : 200003D8
CPUID: 410FC241
HFSR : 40000000
MMFSR: 1
BFSR : 0
UFSR : 0
DFSR : 0
AFSR : 0
Mode : Thread
Priv : Privileged
Stack: PSP

-- MbedOS Fault Handler --



++ MbedOS Error Info ++
Error Status: 0x80FF013D Code: 317 Module: 255
Error Message: Fault exception
Location: 0x1C6822D0
Error Value: 0x20000600
Current Thread: application_unnamed_thread Id: 0x20011090 Entry: 0x144BD StackSize: 0x4000 StackMem: 0x20011898 SP: 0x20015788
For more info, visit: https://mbed.com/s/error?error=0x80FF013D&tgt=NUMAKER_PFM_M487
-- MbedOS Error Info --

= System will be rebooted due to a fatal error =
= Reboot count(=1) reached maximum, system will halt af
```

If I patch this line:

https://github.com/Azure/azure-iot-sdk-c/blob/2bff372a9b22b54d3d4bc9a3e74e349f68fcde46/iothub_client/src/iothubtransport_mqtt_common.c#L3228-L3229

To:

```C
            if (transport_data->xioTransport != NULL) {
                mqtt_client_dowork(transport_data->mqttClient);
            }
```

I meet `IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN` error but no crash:

```
PHY ID 1:0x243
PHY ID 2:0xc54
Info: Connection success, MAC: 96:38:0:53:80:5c
Info: Getting time from the NTP server
Info: Time: Fri Sep25 3:27:1 2020

Info: RTC reports Fri Sep25 3:27:1 2020

Info: Starting the Demo
Info: Initializing IoT Hub client
Info: Sending: "10 messages left to send, or until we receive a reply"
-> 3:27:2 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: nuvoton-test-001.azure-devices.net/my-symm-device-001/?api-version=2019-10-01&DeviceClientType=iothubclient%2f1.3.9%20(native%3b%20mbedOS5%3b%20undefined) | PWD: XXXX | CLEAN: 0
<- 3:27:2 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
Info: Connected to IoT Hub
-> 3:27:2 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: devices/my-symm-device-001/messages/devicebound/# | QOS: 1
-> 3:27:3 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 3 | PAYLOAD_LEN: 53
<- 3:27:3 SUBACK | PACKET_ID: 2 | RETURN_CODE: 1
Info: Sending: "9 messages left to send, or until we receive a reply"
-> 3:27:3 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 4 | PAYLOAD_LEN: 52
<- 3:27:3 PUBACK | PACKET_ID: 3
Info: Message sent successfully
<- 3:27:3 PUBACK | PACKET_ID: 4
Info: Message sent successfully
Info: Sending: "8 messages left to send, or until we receive a reply"
-> 3:27:4 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 5 | PAYLOAD_LEN: 52
<- 3:27:4 PUBACK | PACKET_ID: 5
Info: Message sent successfully
Info: Sending: "7 messages left to send, or until we receive a reply"
-> 3:27:5 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 6 | PAYLOAD_LEN: 52
<- 3:27:5 PUBACK | PACKET_ID: 6
Info: Message sent successfully
Info: Sending: "6 messages left to send, or until we receive a reply"
-> 3:27:6 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 7 | PAYLOAD_LEN: 52
<- 3:27:6 PUBACK | PACKET_ID: 7
Info: Message sent successfully
Info: Sending: "5 messages left to send, or until we receive a reply"
-> 3:27:7 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 8 | PAYLOAD_LEN: 52
<- 3:27:7 PUBACK | PACKET_ID: 8
Info: Message sent successfully
Info: Sending: "4 messages left to send, or until we receive a reply"
-> 3:27:8 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 9 | PAYLOAD_LEN: 52
<- 3:27:8 PUBACK | PACKET_ID: 9
Info: Message sent successfully
Info: Sending: "3 messages left to send, or until we receive a reply"
-> 3:27:9 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 10 | PAYLOAD_LEN: 52
<- 3:27:9 PUBACK | PACKET_ID: 10
Info: Message sent successfully
Info: Sending: "2 messages left to send, or until we receive a reply"
-> 3:27:10 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 11 | PAYLOAD_LEN: 52
<- 3:27:10 PUBACK | PACKET_ID: 11
Info: Message sent successfully
Info: Sending: "1 messages left to send, or until we receive a reply"
-> 3:27:11 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/my-symm-device-001/messages/events/ | PACKET_ID: 12 | PAYLOAD_LEN: 52
<- 3:27:11 PUBACK | PACKET_ID: 12
Info: Message sent successfully
-> 3:27:13 DISCONNECT
Error: Time:Fri Sep25 3:27:15 2020
 File:.\main.cpp Func:on_connection_status Line:37 Connection failed, reason: IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN
Error: Time:Fri Sep25 3:27:17 2020
 File:.\mbed-client-for-azure\dependencies\azure-umqtt-c\src\mqtt_client.c Func:onOpenComplete Line:454 Error: failure opening connection to endpoint
Error: Time:Fri Sep25 3:27:30 2020
 File:.\mbed-client-for-azure\azure-iot-sdk-c\iothub_client\src\iothubtransport_mqtt_common.c Func:InitializeConnection Line:2424 mqtt_client timed out waiting for CONNACK
-> 3:27:31 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: nuvoton-test-001.azure-devices.net/my-symm-device-001/?api-version=2019-10-01&DeviceClientType=iothubclient%2f1.3.9%20(native%3b%20mbedOS5%3b%20undefined) | PWD: XXXX | CLEAN: 0
<- 3:27:32 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
Info: Connected to IoT Hub
-> 3:27:32 SUBSCRIBE | PACKET_ID: 13 | TOPIC_NAME: devices/my-symm-device-001/messages/devicebound/# | QOS: 1
<- 3:27:32 SUBACK | PACKET_ID: 13 | RETURN_CODE: 1
```

**Comment by danewalton** [comment]
Interesting thanks for the logs. I tried repro'ing this scenario on our Linux/Windows samples and don't seem to be running into the same issue. Let me check into it further to see if there's something specific about the mbed case which could be causing this.

**Comment by danewalton** [comment]
Hi @ccli8 thanks for the investigative work. I think I have an idea what might be going on here. I'll try and explain the flow as I see it (pardon the length here).

**Calling `IoTHubTransport_MQTT_Common_DoWork`**

The input handle gets cast to `PMQTTTRANSPORT_HANDLE_DATA transport_data`. Inside this struct, there lives:
https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L193-L194
Inside of the above member `mqtt_client` there lives:
https://github.com/Azure/azure-umqtt-c/blob/10aac20a4e66155183ebb1706f4c91a600dcafdd/src/mqtt_client.c#L47

AKA there exists the following three things:
```c
PMQTTTRANSPORT_HANDLE_DATA transport_data;
transport_data.mqttClient; // MQTT_CLIENT_HANDLE
transport_data.xioTransport; // XIO_HANDLE
transport_data.mqttClient.xioHandle; // XIO_HANDLE
```

As you can see we have two references to an `XIO`. Now the `XIO` gets bound to the mqttClient here:
https://github.com/Azure/azure-umqtt-c/blob/10aac20a4e66155183ebb1706f4c91a600dcafdd/src/mqtt_client.c#L1028
and for the `PMQTTTRANSPORT_HANDLE_DATA ` here:
https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L2159

The problem comes when we call `if (InitializeConnection(transport_data) == 0)` on line 3174 here:
https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L3174

**Calling `InitializeConnection(transport_data)`**

In here, on line 2444, we can try and execute a Disconnect:
https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L2444-L2449
In that disconnect, we issue the disconnect packet and then, *crucially and fatally*, we call the `DestroyXioTransport` on line 1897 here:
https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L1897
That frees and sets to `NULL` the xio stored in `PMQTTTRANSPORT_HANDLE_DATA`. However it DOES NOT free and set to `NULL` the `XIO` stored in `transport_data.mqttClient` (IE `transport_data.mqttClient.xioHandle`).

Later, when we kick out of the `InitializeConnection` call, [it returns success (0)](https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L2448) and we can make our way down to [line 3266](https://github.com/Azure/azure-iot-sdk-c/blob/ff1393162fe5624d02abe6deabc3509ee6d9b08e/iothub_client/src/iothubtransport_mqtt_common.c#L3266) where we invoke the `mqtt_client_dowork()` passing the `transport_data.mqttClient` which does not have a `NULL`ed out xio transport.

Seems like we have a bug! Thanks again for the work! I'll start working on a fix here and post updates as it progresses.

**Comment by ccli8** [comment]
Thanks for your detailed explanation. Yes, this is just the flow I met.

**Comment by danewalton** [comment]
Okay I luckily have a board that can run this demo. I was finally able to get a repro but only on the board. It's odd that the scenario which I created to repro the issue it does not happen on the Linux/Windows version. I'll keep digging to see what's up and get back to you.

**Comment by danewalton** [comment]
Alrighty @ccli8 I believe I understand what is going on here. Bear with me with the explanation.

The gist: there is a problem with the `socketio_mbed_os5.cpp` implementation.

**From The Top**
What we want is for the xio transport to be deleted in the `mqtt_client` (aka set it to `NULL`). That happens here:
https://github.com/Azure/azure-umqtt-c/blob/10aac20a4e66155183ebb1706f4c91a600dcafdd/src/mqtt_client.c#L111
which gets called from here:
https://github.com/Azure/azure-umqtt-c/blob/10aac20a4e66155183ebb1706f4c91a600dcafdd/src/mqtt_client.c#L1289
(pasting the snippet here)
```c
if (mqtt_client->mqtt_status & MQTT_STATUS_PENDING_CLOSE)
{
    close_connection(mqtt_client);
    // turn off pending close
    mqtt_client->mqtt_status &= ~MQTT_STATUS_PENDING_CLOSE;
}
```

So what we seem to be missing is `mqtt_client->mqtt_status` getting the `MQTT_STATUS_PENDING_CLOSE` flag turned on.

If we look for all the cases when we set that (`|= MQTT_STATUS_PENDING_CLOSE`), there are three places which it occurs, but only one we care about which is here:
https://github.com/Azure/azure-umqtt-c/blob/10aac20a4e66155183ebb1706f4c91a600dcafdd/src/mqtt_client.c#L220

I had set a breakpoint on the board to see when that callback was invoked but it turns out, while the Linux/Windows version had consistent hits, on the board, it never happened. So diving into the Linux/Windows version, it eventually gets invoked in the socket_io here:
https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_berkeley.c#L964

If we look at the socket_io implementation for the `socketio_mbed_os5.cpp`, it looks like this:
```c
int socketio_send(CONCRETE_IO_HANDLE socket_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
{
    int result;

    if ((socket_io == NULL) ||
        (buffer == NULL) ||
        (size == 0))
    {
        /* Invalid arguments */
        result = MU_FAILURE;
    }
    else
    {
        result = 0;

        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
        if (socket_io_instance->io_state != IO_STATE_OPEN)
        {
            result = MU_FAILURE;
        }
        else
        {
            // Queue the data, and the socketio_dowork sends the package later
            if (add_pending_io(socket_io_instance, (const unsigned char*) buffer, size, on_send_complete, callback_context) != 0)
            {
                result = MU_FAILURE;
            }
        }
    }

    return result;
}
```
So the callback is never invoked. I tried adding the callback invocation so that it looks like this:
```c
int socketio_send(CONCRETE_IO_HANDLE socket_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
{
    int result;

    if ((socket_io == NULL) ||
        (buffer == NULL) ||
        (size == 0))
    {
        /* Invalid arguments */
        result = MU_FAILURE;
    }
    else
    {
        result = 0;

        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
        if (socket_io_instance->io_state != IO_STATE_OPEN)
        {
            result = MU_FAILURE;
        }
        else
        {
            // Queue the data, and the socketio_dowork sends the package later
            if (add_pending_io(socket_io_instance, (const unsigned char*) buffer, size, on_send_complete, callback_context) != 0)
            {
                result = MU_FAILURE;
            }
            // ADDITION HERE
            if (on_send_complete != NULL)
            {
                on_send_complete(callback_context, IO_SEND_OK);
            }
        }
    }

    return result;
}
```

Once I did that, the hardfault went away.

To be clear, I don't know what possible other ramifications there are for putting that there since technically, the mbed version of socket IO doesn't send the packets in `socketio_send` but instead adds them as "pending", which eventually get sent in the `socketio_dowork`. But I will say that the sample will run from start to finish without a hardfault.

Please feel free to give that a try and let me know if that works.

**Comment by ccli8** [comment]
@danewalton Many thanks for your effort. The hard fault also disappears with the patch per my test, but the patch is temporary and doesn't get to the point.

Even in Linux/Windows `socketio_send` implementation, the real send work can delay to `socketio_dowork` on `EAGAIN` error, so the same crash can still happen?

https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_berkeley.c#L949

Now that delay send through `add_pending_io` is valid, there's inconsistency with time to destroy `xio` transport: `iothubtransport_mqtt_common.c` destroys `xio` transport on `socketio_send` return immediately but `mqtt_client.c` delays the destroy (set to `null`) to `on_send_complete` callback possibly in delayed send.

**Comment by danewalton** [comment]
Yes this is definitely worth looking into agreed. I'll see if I can investigate and get some answers here.

**Comment by danewalton** [comment]
Okay here's what I have found after running it on Windows. (I'll try to copy and link code snippets that are relevant).

Let's assume that we are [here](https://github.com/Azure/azure-iot-sdk-c/blob/3871e2fa70de79f72720bc4fdc21e4f3251d0269/iothub_client/src/iothubtransport_mqtt_common.c#L1888) calling `mqtt_client_disconnect()` from `DisconnectFromClient()`. As I had written before, this propagates down to the `socket_io` which brings us to [here](https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_win32.c#L573-L606):
```c
int send_result = send(socket_io_instance->socket, (const char*)buffer, (int)size, 0);
if (send_result != (int)size)
{
    int last_error = WSAGetLastError();

    if (send_result == SOCKET_ERROR && last_error != WSAEWOULDBLOCK)
    {
        LogError("Failure: sending socket failed %d.", last_error);
        result = MU_FAILURE;
    }
    else
    {
        size_t bytes_sent = (send_result == SOCKET_ERROR ? 0 : send_result);

        /* queue data */
        if (add_pending_io(socket_io_instance, ((const unsigned char*)buffer) + bytes_sent, size - bytes_sent, on_send_complete, callback_context) != 0)
        {
            LogError("Failure: add_pending_io failed.");
            result = MU_FAILURE;
        }
        else
        {
            result = 0;
        }
    }
}
else
{
    if (on_send_complete != NULL)
    {
        on_send_complete(callback_context, IO_SEND_OK);
    }

    result = 0;
}
```
Should there be an `EAGAIN` (or here `WSAEWOULDBLOCK`), the data is queued up via `add_pending_io()` and the `on_send_complete()` callback does not get invoked as you correctly stated. So this kicks all the way up to return from our original call to `mqtt_client_disconnect()` from `DisconnectFromClient()`.

This now puts us in the do-while loop [here](https://github.com/Azure/azure-iot-sdk-c/blob/3871e2fa70de79f72720bc4fdc21e4f3251d0269/iothub_client/src/iothubtransport_mqtt_common.c#L1890-L1895):
```c
size_t disconnect_ctr = 0;
do
{
    mqtt_client_dowork(transport_data->mqttClient);
    disconnect_ctr++;
    ThreadAPI_Sleep(50);
} while ((disconnect_ctr < MAX_DISCONNECT_VALUE) && (transport_data->disconnect_recv_flag == 0));
```

This will loop until the `disconnect_recv_flag` gets set to `1` which happens in the `mqtt_disconnect_cb(void* ctx)` [here](https://github.com/Azure/azure-iot-sdk-c/blob/3871e2fa70de79f72720bc4fdc21e4f3251d0269/iothub_client/src/iothubtransport_mqtt_common.c#L1866-L1873).

The call stack for that invocation looks like this:
![image](https://user-images.githubusercontent.com/36710865/95268024-5b92fa00-07eb-11eb-8b66-4e01675015b4.png)
which means it will only be set once the socket is closed.

Otherwise, the code will sit in that do-while and not move on to the `DestroyXioTransport(transport_data);` which comes after. With that, I don't think there is an issue here. Please absolutely feel free to respond with something you think I might be overlooking or something you find yourself.

**Comment by ccli8** [comment]
Following your finding, I look into the `do-while` loop and find in my case timeout (`MAX_DISCONNECT_VALUE`) happens, so `DestroyXioTransport` is called with `disconnect_recv_flag` not set. If I enlarge `MAX_DISCONNECT_VALUE` to 500 from 50, I find `disconnect_recv_flag` finally gets set but it may take 15s (compared to originally 2.5s). No crash after this modification. Why it takes so long may be associated with `IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN` error, but it is another story for my case. Timeout, caused by bad network or other error can cause `DestroyXioTransport` called with `disconnect_recv_flag` unset yet, so the inconsistency on time to destroy `xio` can still exist.

```C++
size_t disconnect_ctr = 0;
do
{
    mqtt_client_dowork(transport_data->mqttClient);
    disconnect_ctr++;
    ThreadAPI_Sleep(50);
} while ((disconnect_ctr < MAX_DISCONNECT_VALUE) && (transport_data->disconnect_recv_flag == 0));
```

**Comment by danewalton** [comment]
It's possible that could also be due to the lack of the callback in the mbed_os5 socket_io as well (in the `socketio_dowork()` function). I can't confirm right now but preliminary testing in Linux/Windows doesn't have the hard fault should `disconnect_ctr` expire. I will do some more testing though.

**Comment by danewalton** [comment]
Hi @ccli8 I think I have an update here.

So here's where I think we have an issue (probably beating a dead horse with you here but [link to code for reference](https://github.com/Azure/azure-iot-sdk-c/blob/3871e2fa70de79f72720bc4fdc21e4f3251d0269/iothub_client/src/iothubtransport_mqtt_common.c#L1875-L1903)):

![image](https://user-images.githubusercontent.com/36710865/96016747-f9934f80-0dfd-11eb-9f4a-61cee1a8ed6c.png)

As before I force an expired SAS token. I first break at line 1888 and then break [here](https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_berkeley.c#L935) in the socket_io layer at the call to send(). I force the return value to be EGAIN which adds it to the queue by add_pending_io(). It kicks all the way back up to the picture above at line 1892 to mqtt_client_dowork() and I break again [here](https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_berkeley.c#L996) in the socketio_dowork(). Here, I again force the errno to be EAGAIN so it breaks out of the while loop on [line 1004](https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_berkeley.c#L1004). After forcing errno to be [EAGAIN on recv to avoid an indicate_error()](https://github.com/Azure/azure-c-shared-utility/blob/ef6ebf91a2e3530275050f0f789bab9f744c45e9/adapters/socketio_berkeley.c#L1048-L1066), the code then kicks all the way back up to the above picture at line 1893. Assuming this behavior continued for the duration of MAX_DISCONNECT_VALUE, it breaks out of the loop with mqttClient->xiotransport not being set to NULL but the underlying xio being destroyed on line 1897. Code then progresses [here](https://github.com/Azure/azure-iot-sdk-c/blob/3871e2fa70de79f72720bc4fdc21e4f3251d0269/iothub_client/src/iothubtransport_mqtt_common.c#L3266) where the mqtt_client do work is invoked with the reference to xio valid, and it hard faults.

I think I have a fix which I will post shortly.

**Comment by ccli8** [comment]
I've confirmed I meet SAS expire error `IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN` first and then run into the crash path discussed above. After I fix SAS expire error, I won't go into the crash path. Anyway, the crash path above is a potential issue, and happy that you can deliberately re-generate it now.

**Comment by az-iot-builder-01** [comment]
@danewalton, @ccli8, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by danewalton** [comment]
Reopen b/c the pr mistakenly closed it

**Comment by danewalton** [comment]
Hi @ccli8 thanks again for all the work you put in. I merged the fix which should clear that reference. Please let me know if you see the problem go away with this fix.
Thanks,
Dane

**Comment by ccli8** [comment]
@danewalton Yes, the issue gets fixed on master (2dc88db7d4acf9dbeeacd727d58fd62f9925c9b4). Many thanks for your support.

**Comment by danewalton** [comment]
@ccli8 no thank you. you did great investigative work and made replicating the problem a lot easier. don't hesitate to let us know any other hiccups you find!
